File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.length();
        int left = 0 , right = 0 , maxLen = 0;
        unordered_set<char> window;
        while (right < n) {
            if (window.find(s[right]) == window.end()) {
                window.insert(s[right]);
                maxLen = max(maxLen, right-left+1);
                right++;
            }
            else {
                window.erase(s[left]);
                left++;
            }
        }
        return maxLen;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int left = 0;
        int right = 0;
        int count = 0;
        int freq[] = new int[128];
        int maxi = 0;
        while (right < n) {
            char ch = s.charAt(right);
            freq[ch]++;
            count++;
            while (freq[ch] > 1) {
                freq[s.charAt(left)]--;
                count--;
                left++;
            }
            maxi = Math.max(maxi, count);
            right++;
        }
        return maxi;
    }
}